### PR TITLE
Refactor SPI interfaces - use single common ConfigExtension to ServiceLoad all avaje-config extensions

### PR DESCRIPTION
**Pull request by SentryMan**
This is a breaking change for avaje-config extensions.

Makes one super interface `ConfigExtension` for all the spi classes. All avaje-config extensions need to register with ServiceLoader using this interface rather than the specific one (like ConfigurationSource).

All these extensions are loaded via ServiceLoader via the `ConfigExtension` interface rather than making the 6 separate service loader calls for each specific extension interfaces.

Requires https://github.com/avaje/avaje-spi-service/pull/18

### CODE REVIEW

**Comment by rob-bygrave** [comment]
> moves pre/post initialization from the Config Log into the Config Plugin

I don't think this is what we want. Those methods were specifically on ConfigurationLog for a reason so its unclear why you moved them?  If we where to make such a change we should communicate it clearly (and as such our github issue title should reflect all the change included in a PR).

Why did you want to include this change?

File: avaje-config/src/main/java/io/avaje/config/CoreConfigurationBuilder.java [hunk 1]
```diff
@@ -1,35 +1,56 @@
...
+  private ResourceLoader resourceLoader = spiResourceLoader;
+  private ModificationEventRunner eventRunner = spiEventRunner;
   private boolean includeResourceLoading;
   private InitialLoader initialLoader;
 
-  private static ConfigurationLog initialiseLog() {
-    return ServiceLoader.load(ConfigurationLog.class)
-        .findFirst()
-        .orElseGet(DefaultConfigurationLog::new);
+  static {
```
Comment (rob-bygrave) [inline avaje-config/src/main/java/io/avaje/config/CoreConfigurationBuilder.java:37]: I was expecting for there to be a separate class that owned all the service loading. Hmmm.

File: avaje-config/src/main/java/io/avaje/config/Parsers.java [hunk 2]
```diff
@@ -35,11 +35,16 @@ private void initYamlParser() {
   }
 
   private void initParsers() {
-    ServiceLoader.load(ConfigParser.class).forEach(p -> {
-      for (var ext : p.supportedExtensions()) {
-        parserMap.put(ext, p);
-      }
-    });
+    ServiceLoader.load(ConfigSPI.class)
```
Comment (rob-bygrave) [inline avaje-config/src/main/java/io/avaje/config/Parsers.java:38]: Hmm. I was expecting this code to live with all ServiceLoader code in a specific class dedicated to that job. Pondering ...

File: avaje-config/src/main/java/io/avaje/config/ConfigurationLog.java [hunk 3]
```diff
@@ -9,21 +9,7 @@
  * control how the events are logged. For example, it might delay logging messages
  * until logging implementation has finished configuration.
  */
-public interface ConfigurationLog {
-
-  /**
-   * Invoked when the configuration is being initialised.
-   */
-  default void preInitialisation() {
```
Comment (rob-bygrave) [inline avaje-config/src/main/java/io/avaje/config/ConfigurationLog.java:17]: Removing these methods is a breaking change. I don't understand why this is removed.  Prefer this to be in a separate PR with some justification for the change.

File: avaje-config/src/main/java/io/avaje/config/Parsers.java (hunk 2 above)
Comment (SentryMan) [reply avaje-config/src/main/java/io/avaje/config/Parsers.java:38]: Service Loader is memoized so I thought was fine
